### PR TITLE
HTTP2: Add a property-based test

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR is a proof of concept demonstrating [property-based testing](http://www.aaronstannard.com/fscheck-property-testing-csharp-part1/) for the HPack implementation. It converts the existing HPack roundtrip unit test into a property-based test. Please do not merge as it introduces an unsigned nuget dependency to the http unit tests projects.

### CODE REVIEW

File: src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs [hunk 1]
```diff
@@ -9,27 +9,17 @@
 using System.Net.Http.HPack;
 using System.Net.Http.Headers;
 using System.Text;
+using System.Web;
```
Comment (davidsh) [inline src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs]: This is bringing in System.Web for `HttpUtility` class?

Why not use System.Net.WebUtility instead?

Comment (eiriktsarpalis) [reply src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs]: fixed

**Comment by davidsh** [comment]
@dotnet/ncl Please comment on this PR. What do you think of this new approach for testing?

**Comment by geoffkizer** [comment]
I'm not sure I get it entirely.

It looks like part of the goal is to generate random header names/values for testing. If so, I think that's a fine thing to do, as long as the resulting values are deterministic and reproducible.

That said, it seems like we could do this with a lot less added scaffolding. I'm not sure what to make of the [Property] stuff and all the Gen<T> stuff etc. Seems like we should be able to do accomplish the same thing within the existing Theory/MemberData framework by just writing a small amount of code to generate some pseudorandom cases.

**Comment by eiriktsarpalis** [comment]
> It looks like part of the goal is to generate random header names/values for testing. If so, I think that's a fine thing to do, as long as the resulting values are deterministic and reproducible.

Correct, this will run the unit test with 100 pseudo-random inputs. The run can be made reproducible by fixing a seed on the attribute. Alternatively, we could put the test in the outer loop and let in run in a randomized way.

> That said, it seems like we could do this with a lot less added scaffolding. I'm not sure what to make of the [Property] stuff and all the Gen stuff etc.

One of the benefits of using Gen types is that they allow [shrinking](https://propertesting.com/book_shrinking.html), in cases where a random input is found to invalidate the test. Another is that it provides easy ways to control the probability distribution of your random values. 

Admittedly the example in this PR is particularly involved, primarily because of constraints in valid http headers and the fact that I wanted to include static table values in the mix. It generally works out of the box for simple values like primitives, strings and arrays.

> Seems like we should be able to do accomplish the same thing within the existing Theory/MemberData framework by just writing a small amount of code to generate some pseudorandom cases.

One issue I have with theories is that it forces the developer to think of test cases a priori. Property-based testing is generally very good at exposing bugs whose existence we cannot necessarily predict. That being said, property-based testing as a technique can be employed to supply inputs for deterministic theories and unit tests.

**Comment by scalablecory** [comment]
I like the ability to generate various permutations of data for testing. I can see it finding bugs similar to what a fuzzer might.

**Comment by scalablecory** [comment]
The randomness is a little scary. I would feel better if I knew this was augmenting specifically-crafted test cases, and that if we ever find a failing case we can add to that list... while still letting the random generator do its job. Presumably this is possible -- can you demonstrate bolting on the existing test cases into this model?

**Comment by scalablecory** [comment]
Is there any way for us to "dial up" the permutation count based on build parameters? It would be nice to have some scheduled runs with more exhaustive permutations.

**Comment by scalablecory** [comment]
I'd love if you can explore a theory-based approach with the same level of permutation creation We've successfully used LINQ or nested for loops to do something like this in various places, so a clear comparison of both approaches would help demonstrate the benefit of adding this extra library.

**Comment by eiriktsarpalis** [comment]
>  I would feel better if I knew this was augmenting specifically-crafted test cases, and that if we ever find a failing case we can add to that list.

Yes, that's the point. A failing counterexample would generally result in a new unit test (or theory input) being added elsewhere. It's not meant to replace traditional unit tests, however it is an excellent tool for guiding development, in a TDD sense.

> Presumably this is possible -- can you demonstrate bolting on the existing test cases into this model?

Not really, but test runs can be made deterministic [by passing a seed parameter](https://github.com/dotnet/corefx/pull/41866/commits/e35819050058ce2b8677c0b5b39daa2ffbc28682).

> Is there any way for us to "dial up" the permutation count based on build parameters? It would be nice to have some scheduled runs with more exhaustive permutations.

Yes, build conditionals should achieve that. I sometimes use higher values for Debug builds, for instance.

> I'd love if you can explore a theory-based approach with the same level of permutation creation. We've successfully used LINQ or nested for loops to do something like this in various places, so a clear comparison of both approaches would help demonstrate the benefit of adding this extra library.

There's no reason why that approach should not be possible in principle, however it has a couple of drawbacks:
* Boilerplate, which increases as you add things like shrinking, controlling the statistical properties of your test.
* Fundamentally, a hand-written generator encodes the bias of the implementer of what expected values ought to be. The power of property-based testing is that, by default, it will attempt to feed your test with any value that is an instance of your input type. It's a blunt instrument meant to test unknown unknowns of your design.

**Comment by scalablecory** [comment]
>> I would feel better if I knew this was augmenting specifically-crafted test cases, and that if we ever find a failing case we can add to that list.
> Yes, that's the point. A failing counterexample would generally result in a new unit test (or theory input) being added elsewhere. It's not meant to replace traditional unit tests, however it is an excellent tool for guiding development, in a TDD sense.
>> Presumably this is possible -- can you demonstrate bolting on the existing test cases into this model?
>Not really, but test runs can be made deterministic by passing a seed parameter.

Okay, so this means the current test code won't go away then -- we need to guarantee that the permutations we specifically guarded against are tested. Which means property-based testing would be extra, applied only to places we want extra fuzzer-like coverage, like perhaps with parsers. That puts this in a more usable light, I think.

**Comment by scalablecory** [comment]
> > I'd love if you can explore a theory-based approach with the same level of permutation creation. We've successfully used LINQ or nested for loops to do something like this in various places, so a clear comparison of both approaches would help demonstrate the benefit of adding this extra library.
> There's no reason why that approach should not be possible in principle, however it has a couple of drawbacks: ...

Using this has its own drawbacks too, right? That's why I'm hoping to see an apples-to-apples comparison to really demonstrate its worth.

We need to consider that learning this library this will be an additional barrier to entry for contribution... which is already a fairly complex process to hurdle. It's worth exploring what can be accomplished using more idiomatic permutation creation. Another angle to consider is what we'd gain by taking the best ideas from FsCheck and add some helper methods into our common testing code.

**Comment by eiriktsarpalis** [comment]
> applied only to places we want extra fuzzer-like coverage, like perhaps with parsers

I would caution that property based testing is very similar to fuzzing but the focus is functional correctness as opposed to security. It's generally excellent for testing small library functions, sorting is a good example.

> Using this has its own drawbacks too, right? That's why I'm hoping to see an apples-to-apples comparison to really demonstrate its worth.

I might have something, almost. HttpStress follows some of the ideas found in property-based testing, namely the randomness of requests can be deterministically replayed using [a seed parameter](https://github.com/dotnet/corefx/blob/46bc2ab4d3ee39f05c431a904a7e8207b16c8890/src/System.Net.Http/tests/StressTests/HttpStress/Program.cs#L53). It was very helpful in identifying an HPack data corruption bug.

> It's worth exploring what can be accomplished using more idiomatic permutation creation. Another angle to consider is what we'd gain by taking the best ideas from FsCheck and add some helper methods into our common testing code.

You might be on to something there. Fundamentally, the key abstraction these libraries expose is the `Gen` type, which essentially has the following shape:
```csharp
interface Gen<T>
{
    IEnumerable<T> Generate(long seed, long size);
}
```
It gives rise to useful reusable combinators (including many LINQ methods). Then, running a property-based test is implementing a runner:
```csharp
interface IRunner
{
    void Run<T>(Predicate<T> test, Gen<T> gen, long seed, long size);
}
```

**Comment by eiriktsarpalis** [comment]
I'm going to close this, since the point of the PR was to trigger a discussion, as opposed to necessarily include the particular change.

**Pull request closed without merging**